### DISCUSSION TITLE
Semantic search and mqtt

### DISCUSSION THREAD

**Discussion by xavan83** [blakeblackshear/frigate#13925] [Ask A Question]
Semantic search is working great and shows impressive results.
I can not see the generated description in the mqtt events.
How can I retrieve it?

**Comment by hawkeye217** [comment]
I've implemented an mqtt topic, `event_update`, that will publish a tracked object ID and description when your AI provider generates a description. https://github.com/blakeblackshear/frigate/pull/13930

**Reply by xavan83** [reply]
Great news.
I try the docker image cffc431 (just the one before you commit) and I receive an error on startup
FileNotFoundError: [Errno 2] No such file or directory: '/config/model_cache/0cb708b4424bfa17362db472a61f7657'

a7ed90f is working fine with the same config file

**Reply by hawkeye217** [reply]
We are making some pretty major changes right now in the dev builds. I'd recommend waiting a bit.

**Reply by divemasterjm** [reply]
i have same error

**Reply by NickM-27** [reply]
we need to see your docker configs

**Reply by divemasterjm** [reply]
here is mine
#version: "3.9"
services:
  frigatelxc:
    container_name: frigatelxc
    privileged: true # this may not be necessary for all setups
    restart: unless-stopped
    network_mode: host
    cap_add:
      - CAP_PERFMON
    image: ghcr.io/blakeblackshear/frigate:61a4a4b #535bf6e 4515eb4 e44a9e8 6bf2708 17fa830 # version 15 18a6aa1 b4acf4f 61854f1 a1fd29b 4f10f82 c17524b  e4ea35e
    shm_size: "4096mb" # update for your cameras based on calculation above
    devices:
      - /dev/bus/usb:/dev/bus/usb # passes the USB Coral, needs to be modified for other versions
on-linux
      - /dev/dri/renderD128 # for intel hwaccel, needs to be updated for your hardware
    volumes:
      - /etc/localtime:/etc/localtime:ro
      - /root/frigate/config.yml:/config/config.yml:rw
      - /root/frigate/go2rtc.yaml:/config/frigate-go2rtc.yaml:rw
      - /media/frigate:/media/frigate
      - /root/frigate/db:/db
      - /root/frigate:/dev/shm/logs
      - type: tmpfs # Optional: 1GB of memory, reduces SSD/SD Card wear
        target: /tmp/cache
        tmpfs:
          size: 2000000000
    ports:
      - "5000:5000"
      - "1935:1935" # RTMP feeds
      - "8554:8554"
      - "8555:8555"
      - "8080:8080"
    environment:
      FRIGATE_RTSP_PASSWORD: "password"
      PLUS_API_KEY: xxxxx
      TZ: Europe/Madrid

**Reply by NickM-27** [reply]
> - /root/frigate/config.yml:/config/config.yml:rw

this is incorrect, you should have the entire config directory mapped not just the config file

**Reply by divemasterjm** [reply]
if i map /root/frigate:/config:rw
if got errors saying go2rtc is a directory
2024-09-24 17:34:24.009826457  [INFO] Starting go2rtc healthcheck service...
2024-09-24 17:34:24.608236186  [INFO] Removing stale config from last run...
2024-09-24 17:34:24.608910695  [INFO] Preparing new go2rtc config...
2024-09-24 17:34:24.755377003  [INFO] Not injecting WebRTC candidates into go2rtc config as it has been set manually
2024-09-24 17:34:24.775491118  [WARN] Using go2rtc binary from '/config/go2rtc' instead of the embedded one
2024-09-24 17:34:24.775494303  [INFO] Starting go2rtc...
2024-09-24 17:34:24.775503523  ./run.user: line 77: /config/go2rtc: Is a directory
2024-09-24 17:34:24.775511375  ./run.user: line 77: exec: /config/go2rtc: cannot execute: Is a directory
2024-09-24 17:34:24.778110487  [INFO] The go2rtc service exited with code 126 (by signal 0)
2024-09-24 17:34:25.782473102  [INFO] Removing stale config from last run...
2024-09-24 17:34:25.783337615  [INFO] Preparing new go2rtc config...
2024-09-24 17:34:25.931971700  [INFO] Not injecting WebRTC candidates into go2rtc config as it has been set manually
2024-09-24 17:34:25.952059129  [WARN] Using go2rtc binary from '/config/go2rtc' instead of the embedded one
2024-09-24 17:34:25.952061399  [INFO] Starting go2rtc...
2024-09-24 17:34:25.952062145  ./run.user: line 77: /config/go2rtc: Is a directory
2024-09-24 17:34:25.952062785  ./run.user: line 77: exec: /config/go2rtc: cannot execute: Is a directory
2024-09-24 17:34:25.954467572  [INFO] The go2rtc service exited with code 126 (by signal 0)

**Reply by NickM-27** [reply]
what are the contents of `/root/frigate`

**Reply by NickM-27** [reply]
if you have a directory called go2rtc there, you should not

**Reply by divemasterjm** [reply]
i delete the directory but when i deploy it creates automaticaly
now i have this mappings
      - /etc/localtime:/etc/localtime:ro
      - /root/frigate:/config:rw
      - /media/frigate:/media/frigate
      - /root/frigate/db:/db
      - /root/frigate:/dev/shm/logs

**Reply by divemasterjm** [reply]
the content of /root/frigate
root@frigate:~/frigate# ls -al
total 4268
drwxrwxrwx  9 root   root       4096 Sep 24 17:41 .
drwx------ 10 root   root       4096 Sep 24 17:19 ..
-rwxrwxrwx  1 root   root         17 Sep 24 17:13 .exports
-rwxrwxrwx  1 root   root        128 Sep 24 17:13 .jwt_secret
-rwxrwxrwx  1 root   root         17 Sep 24 17:13 .timeline
-rwxrwxrwx  1 root   root         16 Sep 24 17:13 .vacuum
drwxr-sr-x  2 nobody nogroup    4096 Sep  7 11:10 certsync
drwxr-sr-x  2 nobody nogroup    4096 Sep 24 17:37 chroma
-rwxrwxrwx  1 root   root      25767 Sep  5 08:13 config.05092024
-rwxrwxrwx  1 root   root      25767 Sep  6 07:11 config.06092024
-rwxrwxrwx  1 root   root      25767 Sep  8 07:42 config.08092024
-rwxrwxrwx  1 root   root      21327 Jun 10 16:34 config.100624
-rwxrwxrwx  1 root   root      23298 Jul 13 09:05 config.bueno
-rwxrwxrwx  1 root   root      24750 Jul 16 19:29 config.buenosingo2rtc.yml
-rwxrwxrwx  1 root   root      26693 Jul 18 07:06 config.congo2rtc
-rwxrwxrwx  1 root   root      25512 Jul 18 06:49 config.singo2rtc
-rwxrwxrwx  1 root   root      26943 Sep 22 14:11 config.yml
-rwxrwxrwx  1 root   root      21509 Feb  7  2023 config1024.yml
-rwxrwxrwx  1 root   root      21600 Feb  9  2023 config640.yml
-rwxrwxrwx  1 root   root      20353 Feb  1  2023 configbeta3.yml
-rwxrwxrwx  1 root   root      20924 Feb  1  2023 configbeta4.yml
-rwxrwxrwx  1 root   root      21253 Feb  1  2023 configbeta5.yml
-rwxrwxrwx  1 root   root      25441 Jul 17 19:30 configbuenamixgo2rtc.yml
-rwxrwxrwx  1 root   root      21506 Feb  8  2023 confighd.yml
-rwxrwxrwx  1 root   root      18443 Feb  9  2023 configmodhd.yml
drwxrwxrwx  2 root   root       4096 Sep 24 17:37 db
-rwxrwxrwx  1 root   root       1405 Jan 13  2023 docker-compose.yml
drwxr-sr-x  2 nobody nogroup    4096 Sep 24 17:37 frigate
-rwxrwxrwx  1 root   root        114 Sep 24 17:32 frigate-go2rtc.yaml
drwxrwxrwx  2 root   root       4096 Jun 28 09:13 frigate-notify
-rwxrwxrwx  1 root   root    3861977 Jun  3 22:23 frigate-notify_0.3.0_linux_amd64.tar.gz
drwxrwxrwx  2 root   root       4096 Sep 24 17:13 model_cache
drwxr-sr-x  2 nobody nogroup    4096 Sep 23 19:54 nginx
-rwxrwxrwx  1 root   root        241 Sep 24 17:13 notifications.pem

**Reply by xavan83** [reply]
> I've implemented an mqtt topic, `event_update`, that will publish a tracked object ID and description when your AI provider generates a description. #13930

I have the regenerate button next to the Save button (however these button are not visible at first load because the description textarea is too big, a scrollbar appears)
![image](https://github.com/user-attachments/assets/55f7ccfc-ad35-4eb0-8d37-05e4635b5892)

The regenerate button leads to a 404error in chrome debugger

**Reply by NickM-27** [reply]
because you have

> - /root/frigate:/dev/shm/logs

probably better to do

- /root/frigate/logs:/dev/shm/logs

**Reply by divemasterjm** [reply]
correct, thanks, sorry for my  mistake

**Comment by divemasterjm** [comment]
when i click button to regenerate if got "failed to call gemini for a new description"

**Reply by xavan83** [reply]
Same for me:
![image](https://github.com/user-attachments/assets/7fddc532-5e0f-4b24-a272-c4f9636f6179)

**Reply by NickM-27** [reply]
fixed in https://github.com/blakeblackshear/frigate/pull/13937

**Reply by xavan83** [reply]
> fixed in #13937

Thank you it is working for 2-3 events. After that I receive a message indicates that it can take some time but nothing appears after several minutes.
Which logger can I activate to débug? Like quota issue may be?

**Reply by hawkeye217** [reply]
I've put up some additional fixes here https://github.com/blakeblackshear/frigate/pull/13940

**Reply by xavan83** [reply]
> I've put up some additional fixes here #13940

Still no luck

**Reply by hawkeye217** [reply]
What AI provider are you using? You could be running up against a quota

**Comment by xavan83** [comment]
`frigate  | 2024-09-25 07:22:25.998281479  192.168.0.9 - - [25/Sep/2024:07:22:25 +0200] "PUT /api/events/1727197426.506687-d8ktqg/description/regenerate HTTP/1.1" 200 104 "http://192.168.0.16:5000/explore" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/129.0.0.0 Safari/537.36" "-"
frigate  | 2024-09-25 07:22:26.258377453  [2024-09-25 07:22:26] frigate.genai.gemini           WARNING : Gemini returned an error: 429 Resource has been exhausted (e.g. check quota).
`

You are right
According to the documentation the limit is 15queries per minute (but that is for the flash model, not the pro I was using)


I made some tests regarding the search in the description. One of the car was label as a Renault Kangoo in the description (that is correct). I try to search with Renault Kangoo but the car does not appears in the result.

Do you plan to enhance this feature like doing calls only for specific region and object type?

**Reply by hawkeye217** [reply]
> Do you plan to enhance this feature like doing calls only for specific region and object type?

That has been implemented in https://github.com/blakeblackshear/frigate/pull/13958

**Reply by xavan83** [reply]
> > Do you plan to enhance this feature like doing calls only for specific region and object type?
> 
> That has been implemented in #13958

Thank you. What about the search accuracy? Gemini found some cars brand but I can not find a way to make the search working to find the brand that is in the description.

**Reply by hawkeye217** [reply]
How are you searching? When you disable "thumbnail" in the "search sources" filter, can you find what you're looking for in the description?

**Reply by xavan83** [reply]
I try  gemini-1.5-flash-002 instead of  gemini-1.5-flash
This works but it adds the following string at the beginning of each responses:

> Here's a label for the image:

**Reply by hawkeye217** [reply]
Experiment with custom prompts and ask the AI to not say that. See the docs: https://deploy-preview-13787--frigate-docs.netlify.app/configuration/genai#custom-prompts

**Reply by xavan83** [reply]
This is what I do:
`object_prompts:
        car: "Label the primary vehicle in these images with just the name of the company if it is a delivery vehicle, or the color make and model surrounded by / and the license plate if visible surrounded by #"`

But only gemini-1.5-flash-002 add "Here's a label for the image:" 
gemini-1.5-flash doesn't do that

**Reply by NickM-27** [reply]
The suggestion is to add in the prompt not to say here is the label

**Reply by xavan83** [reply]
> The suggestion is to add in the prompt not to say here is the label

Of course that works.

I found this feature very powerful. I see I can use it with the snapshot. Using the snapshot I can have a far better accuracy for license plate but only with one frame. I try to combine license plate detection (using frigate +) and the semantic search. 
Can a zone be restricted to license_plate sublabel like this (because it does not seems to work):
`plaque:
        objects: license_plate
        coordinates: 0,1,1,1,1,0.376,0,0.386`

**Reply by NickM-27** [reply]
license plates are attributes of car, you can restrict car to a specific zone and specify license plates for that zone only

**Reply by xavan83** [reply]
Like this:
![image](https://github.com/user-attachments/assets/69297231-6565-47bf-9c1e-955b19a91d5d)

Would it be possible to combine such a search with clip (like for now) + a crop of a license_plate snapshot(instead of a full snapshot)?

**Reply by NickM-27** [reply]
you can't apply a license plate filter here because again the license plate does not get attached to the zone

**Reply by xavan83** [reply]
So how to do "and specify license plates for that zone only" ?

**Reply by NickM-27** [reply]
add a prompt for that zone and make it be "and specify the license plate number if there is one visible"
